groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java

```diff
@@ -125,6 +143,72 @@ public Path getWorkspaceRoot() {
     }
 
     @Override
+    public void handleFileChanged(DidChangeTextDocumentParams params) {
+        if (params.getContentChanges().isEmpty() || !Paths.get(params.getTextDocument().getUri()).toFile().exists()) {
+            return;
+        }
+        try {
+            SourceWriter sourceWriter = null;
+            Path changedFilePath = Paths.get(params.getTextDocument().getUri());
+            if (originalSourceToChangedSource.containsKey(changedFilePath)) {
+                // New change on existing changed source
+                sourceWriter = originalSourceToChangedSource.get(changedFilePath);
+            } else {
+                // New source to switch out
+                FileUtils.deleteDirectory(config.getTargetDirectory());
+                if (!config.getTargetDirectory().mkdir()) {
+                    throw new RuntimeException(String.format("Could not recreate target directory: %s",
+                            config.getTargetDirectory().getAbsolutePath()));
+                }
+                Path newSourcePath = changedFilesRoot.resolve(workspaceRoot.relativize(changedFilePath));
+
+                sourceWriter = SourceWriter.of(changedFilePath, newSourcePath);
+                originalSourceToChangedSource.put(changedFilePath, sourceWriter);
+            }
+            // Apply changes to source writer and recreate compilation unit
+            sourceWriter.applyChanges(params.getContentChanges());
+            unit = new CompilationUnit(config);
+            addAllSourcesToCompilationUnit();
+        } catch (IOException e) {
+            Throwables.propagate(e);
+        }
+    }
+
+    @Override
+    public void handleFileClosed(String absoluteFilename) {
+        Path originalFile = Paths.get(absoluteFilename);
+        if (!originalSourceToChangedSource.containsKey(originalFile)) {
+            return;
+        }
+        Path changedSource = originalSourceToChangedSource.get(originalFile).getDestination();
+        originalSourceToChangedSource.remove(originalFile);
+        // Deleted the changed file
+        if (!changedSource.toFile().delete()) {
+            log.error("Unable to delete file {}", changedSource.toAbsolutePath());
```

Comment:
feel here unable to remove the file is actually a problem. Shall we throw exception instead of just logging it?